opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,418 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_AQUIFERCT_HEADER_INCLUDED
+#define OPM_AQUIFERCT_HEADER_INCLUDED
+
+#include <opm/parser/eclipse/EclipseState/AquiferCT.hpp>
+#include <opm/parser/eclipse/EclipseState/Aquancon.hpp>
+#include <opm/autodiff/BlackoilAquiferModel.hpp>
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/utility/numeric/linearInterpolation.hpp>
+#include <opm/core/props/BlackoilPhases.hpp>
+
+
+#include <opm/material/densead/Math.hpp>
+#include <opm/material/densead/Evaluation.hpp>
+#include <opm/material/fluidstates/BlackOilFluidState.hpp>
+
+#include <string>
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <map>
+#include <cassert>
+
+namespace Opm
+{
+
+    template<typename TypeTag>
+    class AquiferCarterTracy
+    {
+
+        public:
+            typedef BlackoilModelParameters ModelParameters;
+
+            typedef typename GET_PROP_TYPE(TypeTag, Grid) Grid;
+            typedef typename GET_PROP_TYPE(TypeTag, GridView) GridView;
+            typedef typename GET_PROP_TYPE(TypeTag, Simulator) Simulator;
+            typedef typename GET_PROP_TYPE(TypeTag, FluidSystem) FluidSystem;
+            typedef typename GET_PROP_TYPE(TypeTag, Indices) BlackoilIndices;
+            typedef typename GET_PROP_TYPE(TypeTag, IntensiveQuantities) IntensiveQuantities;
+            typedef typename GET_PROP_TYPE(TypeTag, MaterialLaw) MaterialLaw;
+            typedef typename GridView::template Codim<0>::Entity Element;
+            typedef typename GET_PROP_TYPE(TypeTag, ElementContext) ElementContext;
+
+            static const int numEq = BlackoilIndices::numEq;
+            typedef double Scalar;
+
+            typedef DenseAd::Evaluation<double, /*size=*/numEq> Eval;
+            typedef Opm::BlackOilFluidState<Eval, FluidSystem> FluidState;
+
+            static const auto waterCompIdx = FluidSystem::waterCompIdx;
+            static const auto waterPhaseIdx = FluidSystem::waterPhaseIdx;
+
+
+
+            explicit AquiferCarterTracy( const AquiferCT::AQUCT_data& params, const Aquancon::AquanconOutput& connection,
+                                         const Scalar gravity, const Simulator& ebosSimulator                               )
+            :  ebos_simulator_ (ebosSimulator),
+              aquiferID_ (params.aquiferID),
+              inftableID_ (params.inftableID),
+              pvttableID_ (params.pvttableID),
```

Comment:
Please add `AQUCT_data` as a member (or reference if appropriate) to reduce the complexity of the class `AquiferCarterTracy`.  And `params` is a bad name of this argument, how about `aquct_data` or something similar.